adapters/aja/aja.go

```diff
@@ -0,0 +1,131 @@
+package aja
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/http"
+)
+
+type AJAAdapter struct {
+	endpoint string
+}
+
+func (a *AJAAdapter) MakeRequests(bidReq *openrtb.BidRequest, extraInfo *adapters.ExtraRequestInfo) (adapterReqs []*adapters.RequestData, errs []error) {
+	// split imps by tagid
+	tagIDs := []string{}
+	impsByTagID := map[string][]openrtb.Imp{}
+	for _, imp := range bidReq.Imp {
+		extAJA, err := parseExtAJA(imp)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		imp.TagID = extAJA.AdSpotID
+		imp.Ext = nil
+		if _, ok := impsByTagID[imp.TagID]; !ok {
+			tagIDs = append(tagIDs, imp.TagID)
+		}
+		impsByTagID[imp.TagID] = append(impsByTagID[imp.TagID], imp)
+	}
+
+	req := *bidReq
+	for _, tagID := range tagIDs {
+		req.Imp = impsByTagID[tagID]
+		body, err := json.Marshal(req)
```

Comment:
Do we really need `tagIDs := []string{}`? I believe the following changes use less memory and are equivalent. Am I wrong? Also, `bidReq` is already a shallow copy where is safe to modify `bidReq.Imp`.
```
17   func (a *AJAAdapter) MakeRequests(bidReq *openrtb.BidRequest, extraInfo *adapters.ExtraRequestInfo) (adapterReqs []*adapters.RequestData, errs []error) {
18       // split imps by tagid
19 -     tagIDs := []string{}
20       impsByTagID := map[string][]openrtb.Imp{}
21       for _, imp := range bidReq.Imp {
22           extAJA, err := parseExtAJA(imp)
23           if err != nil {
24               errs = append(errs, err)
25               continue
26           }
27           imp.TagID = extAJA.AdSpotID
28           imp.Ext = nil
29 -         if _, ok := impsByTagID[imp.TagID]; !ok {
30 -             tagIDs = append(tagIDs, imp.TagID)
31 -         }
32           impsByTagID[imp.TagID] = append(impsByTagID[imp.TagID], imp)
33       }
34  
35 -     req := *bidReq
36 -     for _, tagID := range tagIDs {
37 -         req.Imp = impsByTagID[tagID]
38 -         body, err := json.Marshal(req)
   +     for _, imps := range impsByTagID {
   +         bidReq.Imp = imps
   +         body, err := json.Marshal(bidReq)
39           if err != nil {
40               errs = append(errs, err)
41               continue
42           }
43           adapterReqs = append(adapterReqs, &adapters.RequestData{
44               Method: "POST",
45               Uri:    a.endpoint,
46               Body:   body,
47           })
48       }
49  
50       return
51   }
adapters/aja/aja.go
```

Reply:
Yes, I believe it needs. `range impsByTagID` order is changed with each execution and this cause to make tests sometimes failure so it need to fix order with `tagIDs` slice.

> bidReq is already a shallow copy 

Where is it copied? If this is removed, the test is failed.

Reply:
@sambaiz, thanks for the feedback. We were wrong, modifying `bidReq.Imp` at this level does change the original request. Thanks for pointing this out